pkg/controller/grouping/group_entity_index.go

```diff
@@ -0,0 +1,719 @@
+
+package grouping
+
+import (
+	"k8s.io/klog"
+	"reflect"
+	"strings"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/vmware-tanzu/antrea/pkg/apis/core/v1alpha2"
+	"github.com/vmware-tanzu/antrea/pkg/controller/types"
+)
+
+// GroupEntityIndex provides methods to query entities that a given group selects and groups that select a given entity.
+// It maintains indexes between groups and entities to make the query efficient. It supports callers to register
+// callbacks that will be called when a specific type of groups' entities are updated.
+type GroupEntityIndex interface {
+	// AddGroup adds or updates a group to the index. The caller can then get entities selected by this group.
+	AddGroup(groupType string, name string, selector *types.GroupSelector)
+	// DeleteGroup deletes a group from the index.
+	DeleteGroup(groupType string, name string)
+	// AddEventHandler registers an eventHandler for the given type of groups. When any Pod/ExternelEntity/Namespace
+	// update affects the given kind of groups, the eventHandler will be called with the affected groups.
+	AddEventHandler(groupType string, handler eventHandler)
+	// GetEntities returns the selected Pods or ExternalEntities for the given group.
+	GetEntities(groupType string, name string) ([]*v1.Pod, []*v1alpha2.ExternalEntity)
+	// GetGroups returns the groups that select the given Pod.
+	GetGroupsForPod(namespace, name string) (map[string][]string, bool)
+	// AddPod adds or updates a Pod to the index. If any existing groups are affected, eventHandlers will be called with
+	// the affected groups.
+	AddPod(pod *v1.Pod)
+	// DeletePod deletes a Pod from the index. If any existing groups are affected, eventHandlers will be called with
+	// the affected groups.
+	DeletePod(pod *v1.Pod)
+	// AddExternalEntity adds or updates an ExternalEntity to the index. If any existing groups are affected,
+	// eventHandlers will be called with the affected groups.
+	AddExternalEntity(ee *v1alpha2.ExternalEntity)
+	// DeleteExternalEntity deletes an ExternalEntity from the index. If any existing groups are affected, eventHandlers
+	// will be called with the affected groups.
+	DeleteExternalEntity(ee *v1alpha2.ExternalEntity)
+	// AddNamespace adds or updates a Namespace to the index. If any existing groups are affected, eventHandlers will be
+	// called with the affected groups.
+	AddNamespace(namespace *v1.Namespace)
+	// DeleteNamespace deletes a Namespace to the index. If any existing groups are affected, eventHandlers will be
+	// called with the affected groups.
+	DeleteNamespace(namespace *v1.Namespace)
+	// Run starts the index.
+	Run(stopCh <-chan struct{})
+}
+
+type eventHandler func(group string)
+
+// entityType is an internal type used to differentiate Pod from ExternalEntity.
+type entityType string
+
+const (
+	podEntityType      entityType = "Pod"
```

Comment:
If it is used internally only, can we use shorter strings like "p" or "e", hashing/comparison etc. can be faster?